service/server.py

```diff
@@ -32,59 +39,111 @@
 
 
-@app.route('/properties/<postcode>/<street_paon_saon>', methods=['GET'])
-def get_tasks(postcode, street_paon_saon):
-    if postcode == 'N1BLT' and street_paon_saon == 'imaginary-street':
-        abort(404)
-    parts = street_paon_saon.upper().split('_')
-    if len(parts) not in [2, 3]:
-        raise ValueError('Could not split combined street, PAON and SAON into '
-                         'respective parts. Expected street_PAON_SAON.')
-
+def get_query_dict(field_vals):
     query_dict = {
-        'postcode': postcode.upper(),
-        'street': parts[0],
-        'paon': parts[1],
+        'postcode': field_vals[0],
+        'street': field_vals[1],
+        'paon': field_vals[2],
     }
-    if len(parts) == 3:
-        query_dict['saon'] = parts[2]
+    if len(field_vals) == 4:
+        query_dict['saon'] = field_vals[3]
+    return query_dict
 
-    query = PPI_QUERY_TMPL.format(get_query_parts(query_dict))
+
+def get_latest_sale(query_dict):
+    kv_tmpl = '  ?addr lrcommon:{} "{}"^^xsd:string.'
+    kv_lines = [kv_tmpl.format(k, v.upper()) for k, v in query_dict.items()]
+
+    query = PPI_QUERY_TMPL.format('\n'.join(kv_lines))
     ppi_url = ppi_api
     resp = requests.post(ppi_url, data={'output': 'json', 'query': query})
 
     sale_list = resp.json()['results']['bindings']
 
     if len(sale_list) == 0:
-        return jsonify({'message': 'No record found.'})
-    elif len(sale_list) > 1:
-        return jsonify({'message': 'More then one record found.'})
+        return {}
+
+    latest_sale = {k: v['value'] for k, v in sale_list[0].items()}
+    return latest_sale
 
-    latest_sale = sale_list[0]
+
+def get_property_address(query_dict):
+    results = (AddressBase.query.
+        filter_by(postcode=query_dict['postcode']).
+        filter_by(thoroughfareName=query_dict['street']).
+        filter(or_(
+            AddressBase.buildingNumber == query_dict['paon'],
+            AddressBase.buildingName == query_dict['paon'])))
+    if 'saon' in query_dict:
+        results = results.filter_by(subBuildingName=query_dict['saon'])
+
+    return results.all()
+
+
+def create_json(address_rec, latest_sale):
+    # try and get buildingNumber, otherwise buildingName
+    paon = address_rec.buildingNumber or address_rec.buildingName.rstrip()
 
     result = {
-        'saon': 'saon goes here',
-        'paon': 'paon goes here',
-        'street': 'street goes here',
-        'town': 'town goes here',
-        'county': 'county goes here',
-        'postcode': 'postcode goes here',
-        'amount': latest_sale['amount']['value'],
-        'date': latest_sale['date']['value'],
+        'saon': address_rec.subBuildingName.rstrip(),
+        'paon': paon,
+        'street': address_rec.thoroughfareName.rstrip(),
+        'town': address_rec.postTown.rstrip(),
+        'county': address_rec.dependentLocality.rstrip(),
+        'postcode': address_rec.postcode.rstrip(),
+        'amount': latest_sale.get('amount', ''),
```

Comment:
Not sure whether we want to have empty string here as the default? :) What would you think about not including the 'amount' and 'date' keys in the result if there is no data? (I'm essentially concerned about being able to make the difference between a property with faulty/incomplete latest_sale data, and a property with no latest_sale data)